docs/Software Engineering Blog/posts/Comparing_Human_and_AI_way_for_Big-String-Set_Intersection_Performance_in_Java_by_JMH.md

```diff
@@ -0,0 +1,115 @@
+# Comparing Human and AI way for Big-String-Set Intersection Performance in Java by JMH
+
+Recently, I'm facing a case on my job, comparing two set of users and find the intersection as the result. But the difficult part is the number of set is quite big (both set could be 10,000+ elements). To provide a better experiment for our user, the speed can not be too slow. So I want to find a relatively quickly way to do the intersection work. 
+
+In this article, I use AI (gemini, chatGPT) and the 'old-school' way, stackoverflow to find several way to do the set intersection. And Use the JMH [Java Microbenchmark Harness](https://github.com/openjdk/jmh) to run all the test and produce report.
+
+## Candidate Intersection Methods
+
+### Gemini way
+
+First candidate is the Gemini 
+
+![gemini.png](resources/BigSetIntersec/gemini.png)
+
+Gemini provide a way to loop on the bigger set (create a new `HashSet`), and use the `smallSet.contains()` as an if-condition to decide whether to collect the element as the result.
+
+we now call it the `loopBigSetIsContainedByNewSmallSet`
+
+### Gemini way (human modified)
+
+But The Gemini way, create a new HashMap at the first line. it seems a redundant behavior, so we provide a refined way to not create the new HashSet and call it the `loopBigSetIsContainedBySmallSet`.
+
+### ChatGPT way
+
+The second competitor is ChatGPT (GPT4o).
+
+![chatGPT.png](resources/BigSetIntersec/chatGPT.png)
+
+ChatGpt write a straightforward function that using native java method `Set.retainAll` to do the trick. we call it the `smallSetRetainAllBigSet`.
+
+---
+
+After the AI handout their solution, let try some old-school way to find some answer. Here I found the [question](https://stackoverflow.com/questions/2851938/efficiently-finding-the-intersection-of-a-variable-number-of-sets-of-strings) had been asked 3 years before, In this question, the accepted answer is also using the `Set.retainAll`. but we still get two more way from other answer. Let's take a look.
+
+### Stackoverflow way1: stream.filter
+
+The first human provide answer is to use the Java 8 feature to open a stream() on the small set, and check if each element is contained by the big set 
+
+![stream.png](resources/BigSetIntersec/stream.png)
+
+So we called this the `streamSmallSetIsContainedByBigSet` and another version as `streamSmallSetIsContainedByNewBigSet` to compare with the Gemini way.
+
+### Stackoverflow way2: guava
+
+The final one way is to use the Google provide utils library, [Guava](https://github.com/google/guava). There is a method `intersection()` in the Utils-Class `Sets`. 
```

Comment:
🚫 **[vale]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>[Vale.Spelling] Did you really mean 'utils'?